application/backend/app/services/training/otx_trainer.py

```diff
@@ -137,34 +159,127 @@ def assign_subsets(self, project_id: UUID) -> None:
+                sampler=SamplerConfig(**val_subset_sampler_cfg_data), **val_subset_cfg_data
+            )
+            test_subset_config = SubsetConfig(
+                sampler=SamplerConfig(**test_subset_sampler_cfg_data), **test_subset_cfg_data
+            )
+            train_subset_config.transforms = TorchVisionTransformLib.generate(train_subset_config)
+            val_subset_config.transforms = TorchVisionTransformLib.generate(val_subset_config)
+            test_subset_config.transforms = TorchVisionTransformLib.generate(test_subset_config)
+
+            # Wrap them into OTXDataset instances
+            otx_task_type = self.__get_otx_task_type_by_task(task)
+            otx_dataset_class = self.__get_otx_dataset_class_by_task_type(otx_task_type)
+            logger.info("Preparing {} instances for each subset", otx_dataset_class.__name__)
+            otx_training_dataset = otx_dataset_class(
+                dm_subset=dm_training_dataset,
+                transforms=train_subset_config.transforms,
+            )
+            otx_validation_dataset = otx_dataset_class(
+                dm_subset=dm_validation_dataset,
+                transforms=val_subset_config.transforms,
+            )
+            otx_testing_dataset = otx_dataset_class(
+                dm_subset=dm_testing_dataset,
+                transforms=test_subset_config.transforms,
+            )
+
+            # Store the dataset as a new revision
+            logger.info("Saving dataset revision to disk")
+            dataset_revision_id = self._dataset_revision_service.save_revision(
+                project_id=project_id, dataset=dm_dataset
+            )
+            logger.info("Dataset revision saved with ID: {}", dataset_revision_id)
+
             return DatasetInfo(
-                training=dm_dataset.filter_by_subset(Subset.TRAINING),
-                validation=dm_dataset.filter_by_subset(Subset.VALIDATION),
-                testing=dm_dataset.filter_by_subset(Subset.TESTING),
-                revision_id=self._dataset_revision_service.save_revision(project_id, dm_dataset),
+                otx_training_dataset=otx_training_dataset,
+                otx_validation_dataset=otx_validation_dataset,
+                otx_testing_dataset=otx_testing_dataset,
+                otx_training_subset_config=train_subset_config,
+                otx_validation_subset_config=val_subset_config,
+                otx_testing_subset_config=test_subset_config,
+                revision_id=dataset_revision_id,
             )
 
     @step("Prepare Model and Training Configuration")
-    def prepare_model(self, training_params: TrainingParams, dataset_revision_id: UUID) -> None:
+    def prepare_model(
+        self, training_params: TrainingParams, dataset_revision_id: UUID, configuration: TrainingConfiguration
+    ) -> None:
         if training_params.project_id is None:
             raise ValueError("Project ID must be provided for model preparation")
```

Comment:
Since `project_id` is already validated in `run` method, probably we can just cast it similar to L166:

```suggestion
        project_id = cast(UUID, training_params.project_id)
```